### DISCUSSION TITLE
further enhancing the create-dev-cluster script.

### DISCUSSION THREAD

**Discussion by obnoxxx** [rook/rook#13195] [Ideas]
While testing the create-dev-cluster test  script, I noticed a few behaviors that I would propose to change:

1. I was surprised that the script considers a pre-existing minikube environment an error and would like to suggest to treat this situation more gracefully:

-  don't treat existing minikube as an error but allow to use it for rook cluster  installation.
- provide options for  re-creating the  minikube environment. 

2. More  broadly,  the script currently treats the rook cluster and the hosting minikube cluster installation as the same operation. I would like to enhance the script to allow for separately managing minikube environments and rook clusters in minikube, I. e. I'd like to see operations for:

-  creating and deleting a minikube cluster
- creating and deleting a rook-ceph cluster in an existing  minikube cluster
- creating a minikube cluster with a hosted rook cluster in one command as provided by the current functionality.

@rkachach  - what do you think?

**Comment by rkachach** [comment]
Hi @obnoxxx 

Thank you for opening this thread and for your contributions. I'll try to answer the different points expressing my personal opinion but let's wait for input from more users to take a final decision :)

> While testing the create-dev-cluster test script, I noticed a few behaviours that I would propose to change:
> 
> 1. I was surprised that the script considers a pre-existing minikube environment an error and would like to suggest to treat this situation more gracefully:
> 

The current logic of the script basically is conservative: in case it finds a minikube profile with the same name it just fails warning the user that a profile already exists. In case the user wants to continue he just have to use the `-f` flag to force the creation.

> * don't treat existing minikube as an error but allow to use it for rook cluster  installation.
> * provide options for  re-creating the  minikube environment.
> 

I don't see a real benefit of re-using an existing environment because creating a new minikube profile is cheap and a quick operation. Besides, the script by default uses `rook` for the profile unless the user want's a new one in that case he can provide it by using `-p <profile-name>`

> 2. More  broadly,  the script currently treats the rook cluster and the hosting minikube cluster installation as the same operation. I would like to enhance the script to allow for separately managing minikube environments and rook clusters in minikube, I. e. I'd like to see operations for:

IMHO providing extra functionality to manage the `minikube profiles` doesn't make sense. For this you can use minikube directly.

> * creating and deleting a minikube cluster

This is better managed by `minikube` directly

> * creating and deleting a rook-ceph cluster in an existing  minikube cluster

Creating a fresh rook cluster is already covered by the script. Deleting a minikube profile IMHO is better managed by `minikube` directly.

> * creating a minikube cluster with a hosted rook cluster in one command as provided by the current functionality.
> 

This is already covered and it's the main purpose of the script.

> @rkachach - what do you think?

**Comment by BlaineEXE** [comment]
I think it would be helpful to define a "goal" for the script. Who uses it? What features does it offer? What notable features are out of scope?

A script like this has a high amount of risk for scope creep the more people are using it. Having a good idea for what script features are applicable and which are out of scope will help keep the script from getting overcomplicated over time.

IMO, Redo's conceptualization of the script as installing minikube and Rook and treating that as a sort of all-in-one unit is a good one. Users often have problems when they uninstall and re-install Rook when they don't clean up correctly. If new devs use this script, having a new minikube each time will prevent issues like that.

**Comment by rkachach** [comment]
I like the idea of defining the goal and requirements, this way we can take better informed decisions. 

My initial idea behind the script was to `provide a simple way for developers to create a new rook cluster running on minikube`. I wrote the script because I found my self repeating again and again the same steps [from the guide](https://rook.io/docs/rook/v1.12/Contributing/development-environment/#minikube). 

The current version of the script allows the developer to:

1. Create a rook cluster on minikube by using as default the profile `rook` (or another profile specifying the `-p <profile-name>`)
2. Provide config to enable `rook` orchestrator
3. Provide config to enable monitoring

Since the creation of a minikube new environment in general is quick and in order to start with a fresh new rook cluster I designed the script to delete any previous minikube env with the same profile (`rook` or the specified by `-p <profile-name>`). This way, in case the user just run the script accidentally it will not delete his existing profile. For that he must enforce the operation so he intentionally ask the script to delete that profile and create a new one. 

The idea behind the script is to keep it simple and with a simple purpose: `create a rook cluster`. IMHO adding any minikube profile handling in the script doesn't make sense as minikube tool already takes care of that. Incorporating the same logic in the script adds no value and will make its logic way more complicated (and hard to maintain). 

IMHO in general this script should provide options to tweak the created cluster (as enabling/disabling features, etc). So users can use it to setup the cluster that suit their needs to reproduce issues or develop new features etc. As we support having multiple clusters (by using the `-p <profile-name>`) we can have different clusters for different purposes living in the same machine.

My 2 cents :)

**Comment by obnoxxx** [comment]
> Hi @obnoxxx
> 
> Thank you for opening this thread and for your contributions. I'll try to answer the different points expressing my personal opinion but let's wait for input from more users to take a final decision :)

Thanks for sharing your thoughts, @rkachach !
I'll reply inline directly to your points,  trying to explain a little better what was my motivation for proposing these things.

> 
> > While testing the create-dev-cluster test script, I noticed a few behaviors that I would propose to change:
> > 
> > 1. I was surprised that the script considers a pre-existing minikube environment an error and would like to suggest to treat this situation more gracefully:
> 
> The current logic of the script basically is conservative: in case it finds a minikube profile with the same name it just fails warning the user that a profile already exists.

It does not warn but exit with an error instead which  is a big difference.

And note that the current code is not so precise as to check for a profile of the exact name but only for a profile name that contains the string "rook". In my case, the existing cluster with profile name `myrook` leads to the error. 





 







 In case the user wants to continue he just have to use the `-f` flag to force the creation.

Indeed. This triggers  the deletion of the existing cluster and creation of a new environment.

To nit-pick on terminology, this behavior is not what I would call conservative but rather destructive as it would destroy the existing env instead of preserving it.

> 
> > * don't treat existing minikube as an error but allow to use it for rook cluster  installation.
> > * provide options for  re-creating the  minikube environment.
> 
> I don't see a real benefit of re-using an existing environment because creating a new minikube profile is cheap and a quick operation.

In contrast, I experienced it to be a rather time consuming operation. Too much so fo me to want to do it frequently ...


And fwiw, best I can tell, there is no minikube operation that would just create (or delete) a profile. The only way to create a new profile is to create a whole new minikube cluster with the desired profile name. similarly, for deleting a profile, you need to delete the whole cluster behind it.


 Besides, the script by default uses `rook` for the profile unless the user wants a new one in that case he can provide it by using `-p <profile-name>`
> 
> > 2. More  broadly,  the script currently treats the rook cluster and the hosting minikube cluster installation as the same operation. I would like to enhance the script to allow for separately managing minikube environments and rook clusters in minikube, I. e. I'd like to see operations for:
> 
> IMHO providing extra functionality to manage the `minikube profiles` doesn't make sense. For this you can use minikube directly.

I am nor proposing to duplicate or replace  minikube functionality but to separate  in the script the functionalities of management of minikube clusters from that of rook clusters hosted in minikube


> > * creating and deleting a minikube cluster
> 
> This is better managed by `minikube` directly

For me, the main point of the script here is to prevent the user from having to execute the same long minikube commandlines again and again to create a cluster suitable for hosting a rook-ceph cluster ... In that light, I'd find it useful to have this  functionality as subcommands of the script to use whenever needed. 

> > * creating and deleting a rook-ceph cluster in an existing  minikube cluster
> 
> Creating a fresh rook cluster is already covered by the script.

true, but not in a pre-existing minikube cluster.

 Deleting a minikube profile IMHO is better managed by `minikube` directly.

agreed, and that is achieved by deleting the minikube cluster altogether which is something I'd criticize about the minikube cli...


> > * creating a minikube cluster with a hosted rook cluster in one command as provided by the current functionality.
> 
> This is already covered and it's the main purpose of the script.

correct and thats what I'm proposing to separate to an extent to provide the users with more options and flexibility.


> > @rkachach - what do you think?



> Hi @obnoxxx
> 
> Thank you for opening this thread and for your contributions. I'll try to answer the different points expressing my personal opinion but let's wait for input from more users to take a final decision :)
> 
> > While testing the create-dev-cluster test script, I noticed a few behaviours that I would propose to change:
> > 
> > 1. I was surprised that the script considers a pre-existing minikube environment an error and would like to suggest to treat this situation more gracefully:
> 
> The current logic of the script basically is conservative: in case it finds a minikube profile with the same name it just fails warning the user that a profile already exists. In case the user wants to continue he just have to use the `-f` flag to force the creation.
> 
> > * don't treat existing minikube as an error but allow to use it for rook cluster  installation.
> > * provide options for  re-creating the  minikube environment.
> 
> I don't see a real benefit of re-using an existing environment because creating a new minikube profile is cheap and a quick operation. Besides, the script by default uses `rook` for the profile unless the user want's a new one in that case he can provide it by using `-p <profile-name>`
> 
> > 2. More  broadly,  the script currently treats the rook cluster and the hosting minikube cluster installation as the same operation. I would like to enhance the script to allow for separately managing minikube environments and rook clusters in minikube, I. e. I'd like to see operations for:
> 
> IMHO providing extra functionality to manage the `minikube profiles` doesn't make sense. For this you can use minikube directly.
> 
> > * creating and deleting a minikube cluster
> 
> This is better managed by `minikube` directly
> 
> > * creating and deleting a rook-ceph cluster in an existing  minikube cluster
> 
> Creating a fresh rook cluster is already covered by the script. Deleting a minikube profile IMHO is better managed by `minikube` directly.
> 
> > * creating a minikube cluster with a hosted rook cluster in one command as provided by the current functionality.
> 
> This is already covered and it's the main purpose of the script.
